_deploy/r/gnoswap/gns/gns.gno

```diff
@@ -93,93 +129,112 @@ func Render(path string) string {
 	}
 }
 
-// Mint mints GNS to the address.
-// Only emission contract can call Mint.
-func Mint(address pusers.AddressOrName) uint64 {
-	common.IsHalted()
-
-	caller := std.PrevRealm().Addr()
-	if caller != consts.EMISSION_ADDR {
-		panic(addDetailToError(
-			errNoPermission,
-			ufmt.Sprintf("only emission contract(%s) can call Mint, called from %s", consts.EMISSION_ADDR, caller.String()),
-		))
+func checkErr(err error) {
+	if err != nil {
+		panic(err.Error())
 	}
+}
 
-	// if not yet initialized, mint 0 amount
-	if initialized == false {
+// calculateAmountToMint calculates the amount of gns to mint
+// It calculates the amount of gns to mint for each halving year for block range.
+// It also handles the left emission amount if the current block range includes halving year end block.
+func calculateAmountToMint(fromHeight, toHeight int64) uint64 {
```

Comment:
Check this suggestion and see if you can apply to make it more intuitive.

```
func calculateAmountToMint(fromHeight, toHeight int64) uint64 {
	totalAmount := uint64(0)

	fromYear := GetHalvingYearByHeight(fromHeight)
	toYear := GetHalvingYearByHeight(toHeight)

	if isEmissionEnded(fromYear) || isEmissionEnded(toYear) {
		return 0
	}

	for year := fromYear; year <= toYear; year++ {
		if isEmissionEnded(year) {
			break
		}

		yearEndHeight := GetHalvingYearBlock(year)
		mintUntilHeight := std.Min(yearEndHeight, toHeight)
		blocks := uint64(mintUntilHeight - fromHeight + 1)

		// calculate mint amount for the year
		singleBlockAmount := GetAmountByHeight(yearEndHeight)
		yearAmount := singleBlockAmount * blocks

		// handle left emission on the last block of the halving year
		if isLastBlockOfHalvingYear(mintUntilHeight) {
			yearAmount += handleLeftEmissionAmount(year, yearAmount)
		}

		totalAmount += yearAmount
		SetHalvingYearMintAmount(year, GetHalvingYearMintAmount(year)+yearAmount)

		fromHeight = mintUntilHeight + 1
	}

	return totalAmount
```

Reply:
inside for loop
```
if isEmissionEnded(year) {
			break
		}
```
I think it can be removed, before executing loop current logic already checks whether emission is ended or not.

Reply:
```
mintUntilHeight := std.Min(yearEndHeight, toHeight)
```

Unfortunately, `std` doesn't have `Min` method to compare numeric value.
But I get the point of that line. I will make min helper then use it.

Reply:
4078cb4c832f373c9ee78bec19891a992a6d7236